instrumentation/github.com/Shopify/sarama/otelsarama/producer.go

```diff
@@ -155,6 +155,7 @@ func WrapAsyncProducer(saramaConfig *sarama.Config, p sarama.AsyncProducer, opts
 				finishProducerSpan(mc.span, 0, 0, nil)
 			}
 		}()
+		defer close(wrapped.input)
```

Comment:
It looks like this is a consumer of input, not the producer.  If that's the case we shouldn't close it here, but where the producer is finished.

Reply:
Meaning that `close(wrapped.input)` should be called in `AsyncClose` and `Close` methods after sending to message to `input` ?

Reply:
I'm not very familiar with this portion of the code, but generally you want to close where you write to a channel, not where it's read.  And, this function only reads from `wrapped.input`.

If anything you might want to be closing `p.Input()`. And yes I would think that would be more appropriate in the `Close()` of the Producer.

Reply:
@MadVikingGod I will try it next week and do my best to test it from different angles. After a quick look, I have a feeling that it will be a more "correct" code. Thanks a lot 👍

Reply:
@MadVikingGod 
I think that the fix provided in this PR is the optimal one.
We have no other way to know when we can close the input channel other than waiting until `sarama` closes the `producer.Successes()` and `producer.Errors()` channels. It is caused mainly by sarama's `AsyncClose()` implementation which "closes at some point".

Here is the "problematic" code:
- https://github.com/Shopify/sarama/blob/master/async_producer.go#L320-L322
- https://github.com/Shopify/sarama/blob/master/async_producer.go#L1054-L1059

Still, I have managed at least to refactor the existing code a little... I hope it is better.

Reply:
That is generally the accepted way of closing these async channels.  It's acceptable to panic if someone writes after `close()`, because that is the contract that channels have.  

What happens if you stick with the previous behavior, but just close the `p.input` in the `Close()` and `AsyncClose()`?

Reply:
It panics in goroutines created by sarama... So it changes the behavior when instrumentation is used vs not applied. I think it is not acceptable. Let me know if you want stack traces (I can provide them in a few hours).

Reply:
@MadVikingGod 
Thanks lot for your comments 👍 ❤️ 

I took another try. Probably I made some bug when closing the input channel.
Now I am closing the input channel when closing. It is a lot better because the previous " fix" was deadlocking in case of calling `p.Close()` more than once!

Is there anything that does not look good?

Reply:
Getting race conditions in sarama when I do `p,Close()` and `p.Input() <- msg` in different goroutines. Makes sense...